WcaOnRails/app/controllers/registrations_controller.rb

```diff
@@ -439,163 +441,246 @@ def register
+      # stripe_intent contains a Stripe::PaymentIntent as per Stripe documentation
+
+      stored_intent = stored_transaction.stripe_payment_intent
+
+      stored_intent.update_status_and_charges(stripe_intent, audit_event) do |charge_transaction|
+        if stored_intent.holder.is_a? Registration # currently, the only holders that we pay for are Registrations.
+          ruby_money = charge_transaction.money_amount
+
+          stored_intent.holder.record_payment(
+            ruby_money.cents,
+            ruby_money.currency.iso_code,
+            charge_transaction,
+            stored_intent.user.id,
+          )
+        end
+      end
+    when StripeWebhookEvent::PAYMENT_INTENT_CANCELED
+      # stripe_intent contains a Stripe::PaymentIntent as per Stripe documentation
+
+      stored_intent = stored_transaction.stripe_payment_intent
+      stored_intent.update_status_and_charges(stripe_intent, audit_event)
+    else
+      logger.info "Unhandled Stripe event type: #{event.type}"
+    end
+
+    head :ok
+  end
+
   def payment_success
-    @competition = competition_from_params
-    flash[:success] = t("registrations.payment_form.payment_successful")
+    registration = Registration.includes(:competition).find(params[:id])
+    @competition = registration.competition
+
+    # Provided by Stripe upon redirect when the "PaymentElement" workflow is completed
+    intent_id = params[:payment_intent]
+    intent_secret = params[:payment_intent_client_secret]
+
+    stored_transaction = StripeTransaction.find_by(stripe_id: intent_id)
+    stored_intent = stored_transaction.stripe_payment_intent
+
+    unless stored_intent.client_secret == intent_secret
+      flash[:error] = t("registrations.payment_form.errors.stripe_secret_invalid")
+      return redirect_to competition_register_path(@competition)
+    end
+
+    # No need to create a new intent here. We can just query the stored intent from Stripe directly.
+    stripe_intent = stored_intent.retrieve_intent
+
+    unless stripe_intent.present?
+      flash[:error] = t("registrations.payment_form.errors.stripe_not_found")
+      return redirect_to competition_register_path(@competition)
+    end
+
+    stored_intent.update_status_and_charges(stripe_intent, current_user) do |charge_transaction|
+      ruby_money = charge_transaction.money_amount
+
+      registration.record_payment(
+        ruby_money.cents,
+        ruby_money.currency.iso_code,
+        charge_transaction,
+        current_user.id,
+      )
+    end
+
+    # Payment Intent lifecycle as per https://stripe.com/docs/payments/intents#intent-statuses
+    case stored_transaction.status
+    when 'succeeded'
+      flash[:success] = t("registrations.payment_form.payment_successful")
+    when 'requires_action'
+      # Customer did not complete the payment
+      # For example, 3DSecure could still be pending.
+      flash[:warning] = t("registrations.payment_form.errors.payment_pending")
```

Comment:
The `flash` displays a small pop-up like snackbar at the top of the screen. You can see it in action when you sign in and you get a "signed in successfully" -- that's implemented as a `flash`.

The interpretation of this specific case was actually c/p'ed from the previous implementation; GitHub marks it as diff because the lines around it changed too much.

Reply:
I don't know when this would realistically ever come up, but better to have it and never use it than to be sorry we deleted it